[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75305
* #75261
* #75304
* __->__ #72353

Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

dzdang committed:
[Quant][bc-breaking] Combined dispatch registration for max_pool2d & quantized_max_pool2d

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][bc-breaking] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][bc-breaking][devs] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085075).

jerryzh168 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -9312,6 +9308,7 @@
   dispatch:
     CPU: max_pool2d_with_indices_out_cpu
     CUDA: max_pool2d_with_indices_out_cuda
+    QuantizedCPU: max_pool2d_with_indices_out_quantized_cpu
```
can we add a test for `max_pool2d_with_indices` in test_quantized_ops.py? existing test is here: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_op.py#L1222

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qpool.cpp:
```diff
@@ -401,6 +401,29 @@ Tensor quantized_max_pool2d(
+  // the output tensor is already computed from quantized_max_pool2d. junk_out is used as a dummy
+  // argument because it is required by max_pool2d_kernel
+  auto junk_out = at::empty(out.sizes(), self.int_repr().options());
+  max_pool2d_kernel(kCPU, junk_out, indices, self.int_repr(), kW, kH, dW, dH, padW, padH, dilationW, dilationH);
```
we should be able to get the quantized output (quint8) based on the output of max_pool2d_kernel (uint8) I think, the output of max_pool2d_kernel will be the int_repr of the actual quantized output, we can make a quantized tensor with `_make_per_tensor_quantized_tensor `

ezyang replied:
Yes, the code as written looks awful, you're running max_pool2d twice to get out the quantized output and then the indices. Terrible. Please do this modification.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qpool.cpp:
```diff
@@ -401,6 +401,29 @@ Tensor quantized_max_pool2d(
 }
 
+std::tuple<at::Tensor &,at::Tensor &> max_pool2d_with_indices_out_quantized_cpu(const at::Tensor & self, at::IntArrayRef kernel_size, at::IntArrayRef stride, at::IntArrayRef padding, at::IntArrayRef dilation, bool ceil_mode, at::Tensor & out, at::Tensor & indices) {
+  out = at::native::quantized_max_pool2d(self, kernel_size, stride, padding, dilation, ceil_mode);
```
oh we can't remove it now since it's still used in other places, but we can avoiding calling it

jerryzh168 commented:
Hi @brianjo @ezyang could you take a look at this PR to see if the dispatch modifications are OK or not?

ezyang commented:
test failures look alarming

ezyang commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -459,7 +459,7 @@ void max_pool2d_kernel_impl(
   switch (input.suggest_memory_format()) {
     case at::MemoryFormat::Contiguous: {
-      AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, input.scalar_type(), "max_pool2d", [&] {
+      AT_DISPATCH_ALL_TYPES_AND(ScalarType::BFloat16, input.scalar_type(), "max_pool2d", [&] {
```
oh?

ezyang replied:
This looks like you will need an OpInfo update

ezyang commented on aten/src/ATen/native/quantized/cpu/qpool.cpp:
```diff
@@ -401,6 +401,29 @@ Tensor quantized_max_pool2d(
+  const int padW = padding.size() == 1 ? padH : safe_downcast<int, int64_t>(padding[1]);
+
+  const int dilationH = safe_downcast<int, int64_t>(dilation[0]);
+  const int dilationW = dilation.size() == 1 ? dilationH : safe_downcast<int, int64_t>(dilation[1]);
```
Sigh... we really need to get structured kernels working for quantized.

ezyang reviewed (changes requested):
don't run max pool2d twice just cuz you need indices lol

dzdang committed:
Update on "[Quant][bc-breaking][devs] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085075).

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085075).

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085075).

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085075).

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34085075).

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Differential Revision: [D34085075](https://our.internmc.facebook.com/intern/diff/D34085075)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -1262,7 +1262,15 @@ def test_max_pool2d(self, X, kernel, stride, dilation, padding, ceil_mode):
+            "nn.functional with indices": torch.nn.functional.max_pool2d_with_indices,
+        }
 
+        for name, op in ops_under_test.items():
```
looks like we are using `ops_under_test` in both loops, do you mean to use `ops_under_test_with_indices` for one of them?

jerryzh168 reviewed (approved):
LGTM

jerryzh168 commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -56,10 +55,11 @@ void cpu_max_pool(
       // compute local max
       int64_t maxindex = ih0 * input_width + iw0;
-      accscalar_t maxval = -std::numeric_limits<accscalar_t>::infinity();
+      accscalar_t maxval = input.is_quantized() ? std::numeric_limits<accscalar_t>::min()
```
should we use lowest? since that is what we use before in the qpool.cpp

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices. In addition, quantized_max_pool2d has
been removed from the frontend, which was previously not used anyways.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][bc-breaking][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -149,15 +149,15 @@ void cpu_max_pool_channels_last(
       // Pass I: init out lane
       iVec index0_vec = iVec(ih0 * input_width + iw0);
-      Vec out_vec = Vec(-std::numeric_limits<scalar_t>::infinity());
+      Vec out_vec = Vec(input.is_quantized() ? std::numeric_limits<scalar_t>::lowest() : -std::numeric_limits<scalar_t>::infinity());
```
I'm afraid the test isn't going to get optimized away. Do you have benchmark numbers here? You could also make the `is_quantized` a boolean template parameter, so it is guaranteed to be statically known.

ezyang reviewed (changes requested):
I think this will slow down preexisting max_pool2d

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

ezyang commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -25,9 +27,8 @@ void cpu_max_pool(
-
-  auto input_data = input.data_ptr<scalar_t>();
-  auto output_data = output.data_ptr<scalar_t>();
+  auto input_data = std::is_integral<scalar_t>::value ? reinterpret_cast<scalar_t*>(input.data_ptr()) : input.data_ptr<scalar_t>();
```
You're trying very hard not to add a template parameter but it is resulting in code that doesn't make sense. If I didn't know anything about quantization, WHY would I expect an integral `scalar_t` to require reinterpret casting from data ptr. It doesn't make any sense! At the very least there should be a comment but it would be better to structure the code in such a way that people would think "oh yeah, that makes sense." In particular, what is the valid `data_ptr` to extract for quantized types, and why can't it be used here?

jerryzh168 replied:
scalar_t here is probably things like uint8, and input.data_ptr() would have type quint8*, they have the same underlying type but we don't use quint8 for computations. can we do input.data_ptr<uint8*>() now btw, if so we can just remove this cast?

jerryzh168 replied:
thanks for the suggestions! yeah we want to keep `quint8` and `uint8` separate since unlike `uint8`, `quint8` in itself doesn't correspond to any values (doesn't have a meaning). We have done the first point you mentioned

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

ezyang commented on aten/src/ATen/native/quantized/cpu/qpool.cpp:
```diff
@@ -401,6 +402,33 @@ Tensor quantized_max_pool2d(
+  const int dilationW = dilation.size() == 1 ? dilationH : safe_downcast<int, int64_t>(dilation[1]);
+
+  max_pool2d_kernel(kCPU, out, indices, self, kW, kH, dW, dH, padW, padH, dilationW, dilationH);
+  set_quantizer_(out, make_per_tensor_affine_quantizer(self.q_scale(), self.q_zero_point(), out.scalar_type()));
```
random question that is probably not the fault of this PR: is it really correct to `set_quantizer_` on the out argument like this? Suppose I take a view of a quantized tensor and then write into it, the quantizers of the view and the base what then become inconsistent. That seems bad!

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

ezyang commented on aten/src/ATen/native/cpu/MaxPoolKernel.cpp:
```diff
@@ -190,12 +199,20 @@ void cpu_max_pool_channels_last(
+              bool res{false};
+              c10::guts::if_constexpr<std::is_integral<scalar_t>::value> (
+                [&res] () { res = false; }, // if integral type
+                [&res, val] () { res = std::isnan(val); }  // if not integral type
```
why not just define a helper function

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).

dzdang committed:
Update on "[Quant][core][improvements] Combined dispatch registration for max_pool2d & quantized_max_pool2d and implemented max_pool2d_with_indices_out_quantized_cpu"


Summary: This PR is part of a series of PRs addressing https://github.com/pytorch/pytorch/issues/54150,
related to using dispatcher for calls to quantized backends as opposed to if/else conditionals.
This particular PR removes the is_quantized check from max_pool2d, and implements a quantized
kernel for max_pool2d_with_indices.

This PR also introduces isnan() support for vectorized int tensors.

This PR relies on https://github.com/pytorch/pytorch/pull/74560, which introduces
structured kernel support for quantized tensors.

Test plan:
```
python test/test_quantization.py -k test_max_pool2d
```

Differential Revision: [D35420901](https://our.internmc.facebook.com/intern/diff/D35420901)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35420901).
